app/controllers/voucher_adjustments_controller.rb

```diff
@@ -51,14 +69,43 @@ def add_voucher
       return false
     end
 
+    # calculate_voucher_adjustment
     clear_payments
 
-    VoucherAdjustmentsService.new(@order).update
-    @order.update_totals_and_states
+    OrderManagement::Order::Updater.new(@order).update_voucher
 
     true
   end
 
+  def load_voucher
+    voucher = Voucher.not_vine.find_by(code: voucher_params[:voucher_code],
+                                       enterprise: @order.distributor)
+    return voucher unless voucher.nil?
+
+    vine_voucher
+  end
+
+  def vine_voucher
+    vine_voucher_validator = Vine::VoucherValidatorService.new(
+      voucher_code: voucher_params[:voucher_code], enterprise: @order.distributor
+    )
+    voucher = vine_voucher_validator.validate
+
+    if vine_voucher_validator.errors.present?
+      return nil if ignored_vine_api_error(vine_voucher_validator)
+
+      @order.errors.add(:voucher_code, I18n.t('checkout.errors.add_voucher_error'))
+      return nil
+    end
+
+    voucher
+  end
+
+  def ignored_vine_api_error(validator)
+    # Ignore distributor not set up for VINE or not found error
+    validator.errors[:vine_settings].present? || validator.errors[:not_found_voucher].present?
+  end
+
   def update_payment_section
     render cable_ready: cable_car.replace(
       selector: "#checkout-payment-methods",
```

Comment:
I didn't replace this by turbo, as it will be done in this PR https://github.com/openfoodfoundation/openfoodnetwork/pull/12896